certora/specs/AccessControlDefaultAdminRules.spec

```
@@ -0,0 +1,484 @@
+import "helpers.spec"
+import "methods/IAccessControlDefaultAdminRules.spec"
+import "methods/IAccessControl.spec"
+import "AccessControl.spec"
+
+use rule onlyGrantCanGrant filtered {
+  f -> f.selector != acceptDefaultAdminTransfer().selector
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Helpers                                                                                                             │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+
+function max_uint48() returns mathint {
+    return (1 << 48) - 1;
+}
+
+function nonZeroAccount(address account) returns bool {
+  return account != 0;
+}
+
+function timeSanity(env e) returns bool {
+  return
+    e.block.timestamp > 0 && // Avoids 0 schedules
+    e.block.timestamp + defaultAdminDelay(e) < max_uint48();
+}
+
+function intervalSanity(env e1, env e2) returns bool {
+  return e1.block.timestamp < e2.block.timestamp;
+}
+
+function delayChangeWaitSanity(env e, uint48 newDelay) returns bool {
+  return e.block.timestamp + delayChangeWait_(e, newDelay) < max_uint48();
+}
+
+function isSet(uint48 schedule) returns bool {
+  return schedule != 0;
+}
+
+function hasPassed(env e, uint48 schedule) returns bool {
+  return schedule < e.block.timestamp;
+}
+
+function min(uint48 a, uint48 b) returns mathint {
+  return a < b ? a : b;
+}
+
+function increasingDelaySchedule(env e, uint48 newDelay) returns mathint {
+  return e.block.timestamp + min(newDelay, defaultAdminDelayIncreaseWait());
+}
+
+function decreasingDelaySchedule(env e, uint48 newDelay) returns mathint {
+  return e.block.timestamp + defaultAdminDelay(e) - newDelay;
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Invariant: defaultAdmin holds the DEFAULT_ADMIN_ROLE                                                                │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant defaultAdminConsistency(address account)
+  defaultAdmin() == account <=> hasRole(DEFAULT_ADMIN_ROLE(), account) 
+  {
+    preserved {
+      // defaultAdmin() returns the zero address when there's no default admin
+      require nonZeroAccount(account);
+    }
+  }
```

Not sure about the preserve here. I'm wondering if we should rather change the invariant like that

```suggestion
invariant defaultAdminConsistency(address account)
  (account == defaultAdmin() && account != 0) <=> hasRole(DEFAULT_ADMIN_ROLE(), account)
```

---

It fails when removing the preserve because `defaultAdmin()` always reports the `address(0)` when there's no admin (eg. has been renounced). So the counterexample is that `defaultAdmin()` can be `address(0)` when `hasRole(DEFAULT_ADMIN_ROLE, address(0)) == false`. If we do your suggestion, we'll end up with a more complex rule:

```
invariant defaultAdminConsistency(address account)
  (account == defaultAdmin() && account != 0 || defaultAdmin() == 0 && account == 0) <=> hasRole(DEFAULT_ADMIN_ROLE(), account)
```

Also, it affects all of the following

```diff
+ assert success <=> hasRole(DEFAULT_ADMIN_ROLE, e.msg.sender)
- assert success <=> e.msg.sender == defaultAdmin()
```

I think is way cleaner keeping the `preserved` since the assumption of non-zero address is reasonable, but I'm open to discussion

---

```
invariant defaultAdminConsistency(address account)
  (account == defaultAdmin() && account != 0 || defaultAdmin() == 0 && account == 0) <=> hasRole(DEFAULT_ADMIN_ROLE(), account)
```
???

if defaultAdmin and account are both 0, `hasRole(DEFAULT_ADMIN_ROLE(), account)` is false.
Also, mixing `&&` and `||` without proper parenthesis is really confusing.

---

Note that in my proposal, 
```(account == defaultAdmin() && account != 0) <=> hasRole(DEFAULT_ADMIN_ROLE(), account)```
if the admin is renounced, and account is 0, the left side is false. I don't see how that is a counterexample

---

The invariant you proposed is still broken for the acceptDefaultAdmin case. What I'm saying is that it ends up in worse readability issues.

We'll keep playing around with this to see if I find a better alternative, but I'd prefer to keep it as it is rn

---

This still doesn't work. I've spent some time trying it out and I don't think it's worth it, the counterexample is well defined and explained, so I'm solving.